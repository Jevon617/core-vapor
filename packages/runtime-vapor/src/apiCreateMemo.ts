import { type Block, type Fragment, fragmentKey } from './apiRender'
import { createComment, createTextNode, insert, remove } from './dom/element'
import { baseWatch } from '@vue/reactivity'
import { extend, hasChanged, isArray } from '@vue/shared'
import { nextTick } from './scheduler'

type BlockFn = (cache?: MemoCache[]) => Block

type MemoCache = {
  memo?: any[]
  watched: boolean
  fragment: Fragment
  shouldUpdate: boolean
}

export const createMemo = (
  memo: () => any[],
  render: BlockFn,
  index: number,
  cache: MemoCache[] = [],
): Fragment => {
  let tempMemo: any[]

  const isNested = index !== 0
  const anchor = __DEV__ ? createComment('memo') : createTextNode()
  const fragment: Fragment = {
    nodes: [],
    anchor,
    [fragmentKey]: true,
  }
  const cached: MemoCache = cache[index] || {
    fragment,
    /**
     * nested createMemo will be called by his parent,
     * so this flag prevents baseWatch from being called repeatedly.
     */
    watched: false,
    /* the memo should cached because the watch's newVal and oldVal are the same array */
    memo: undefined,
    shouldUpdate: false,
  }

  // if nested, the update timing is determined by the parent
  if (isNested) {
    !cached.watched &&
      baseWatch(
        memo,
        (newMemo: any[]) => {
          tempMemo = newMemo.slice()
          handleMemoChange(tempMemo, () => {
            cached.shouldUpdate = true
          })
        },
        { immediate: true, deep: true },
      )
    cached.watched = true
    cached.shouldUpdate && performUpdate(tempMemo!)
    return cached.fragment
  }
  // if not nested, the update timing is determined by the `memo` change
  else {
    baseWatch(
      memo,
      (newMemo: any[]) => {
        tempMemo = newMemo.slice()
        handleMemoChange(tempMemo, () => performUpdate(tempMemo))
      },
      { immediate: true, deep: true },
    )
    return cached.fragment
  }

  function handleMemoChange(newMemo: any[], cb: () => void) {
    const isMount = !isArray(cached.memo)
    if (isMount) {
      cb()
    } else if (!isMemoSame(cached.memo!, newMemo)) {
      nextTick(cb)
    }
  }

  function performUpdate(newMemo: any[]) {
    let { nodes: prevNodes, anchor } = cached.fragment
    let parent = anchor!.parentNode
    let nodes = (cached.fragment.nodes = render(cache))

    if (parent) {
      prevNodes && remove(prevNodes, parent)
      insert(nodes, parent, anchor)
    }
    cache[index] = extend(cached, {
      shouldUpdate: false,
      memo: newMemo,
    })
  }
}

function isMemoSame(preMemo: any[], memo: any[]) {
  if (preMemo.length != memo.length) {
    return false
  }

  for (let i = 0; i < preMemo.length; i++) {
    if (hasChanged(preMemo[i], memo[i])) {
      return false
    }
  }
  return true
}
