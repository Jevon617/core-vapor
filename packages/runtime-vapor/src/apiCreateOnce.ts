import { type Block, type Fragment, fragmentKey } from './apiRender'
import { effectScope } from '@vue/reactivity'
import { createComment, createTextNode } from './dom/element'
import { type ComponentInternalInstance, isVaporComponent } from './component'
import { isArray } from '@vue/shared'
import type { BlockFn } from './apiCreateIf'

/*! #__NO_SIDE_EFFECTS__ */
export const createOnce = (blockFn: BlockFn): Fragment => {
  let scope = effectScope()
  const anchor = __DEV__ ? createComment('once') : createTextNode()
  const fragment: Fragment = {
    nodes: scope.run(() => {
      const block = blockFn()
      const components = getComponents(block)
      components.forEach(comp => comp.scope.stop())
      return block
    })!,
    anchor,
    [fragmentKey]: true,
  }
  scope.stop()
  return fragment
}

function getComponents(block: Block | null) {
  let components: ComponentInternalInstance[] = []
  if (isVaporComponent(block)) {
    components.push(block, ...getComponents(block.block))
  } else if (isArray(block)) {
    block.forEach(child => components.push(...getComponents(child)))
  }
  return components
}
