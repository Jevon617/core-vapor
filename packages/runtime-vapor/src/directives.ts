import { isArray, isFunction } from '@vue/shared'
import {
  type ComponentInternalInstance,
  currentInstance,
  isComponentInstance,
} from './component'
import { pauseTracking, resetTracking, traverse } from '@vue/reactivity'
import { VaporErrorCodes, callWithAsyncErrorHandling } from './errorHandling'
import { renderEffect } from './renderEffect'
import { type Block, fragmentKey } from './apiRender'
import { warn } from './warning'

export type DirectiveModifiers<M extends string = string> = Record<M, boolean>

export interface DirectiveBinding<V = any, M extends string = string> {
  instance: ComponentInternalInstance
  source?: () => V
  value: V
  oldValue: V | null
  arg?: string
  modifiers?: DirectiveModifiers<M>
  dir: ObjectDirective<any, V>
}

export type DirectiveHook<
  T = any | null,
  V = any,
  M extends string = string,
> = (node: T, binding: DirectiveBinding<V, M>) => void

// create node -> `created` -> node operation -> `beforeMount` -> node mounted -> `mounted`
// effect update -> `beforeUpdate` -> node updated -> `updated`
// `beforeUnmount`-> node unmount -> `unmounted`
export type DirectiveHookName =
  | 'created'
  | 'beforeMount'
  | 'mounted'
  | 'beforeUpdate'
  | 'updated'
  | 'beforeUnmount'
  | 'unmounted'
export type ObjectDirective<T = any, V = any, M extends string = string> = {
  [K in DirectiveHookName]?: DirectiveHook<T, V, M> | undefined
} & {
  /** Watch value deeply */
  deep?: boolean
}

export type FunctionDirective<
  T = any,
  V = any,
  M extends string = string,
> = DirectiveHook<T, V, M>

export type Directive<T = any, V = any, M extends string = string> =
  | ObjectDirective<T, V, M>
  | FunctionDirective<T, V, M>

export type DirectiveArguments = Array<
  | [Directive | undefined]
  | [Directive | undefined, () => any]
  | [Directive | undefined, () => any, argument: string]
  | [
      Directive | undefined,
      value: () => any,
      argument: string,
      modifiers: DirectiveModifiers,
    ]
>

export function withDirectives<T extends ComponentInternalInstance | Node>(
  node: T,
  directives: DirectiveArguments,
): T {
  if (!currentInstance) {
    __DEV__ && warn(`withDirectives can only be used inside render functions.`)
    return node
  }

  let realNode: Node

  const instance = currentInstance

  if (isComponentInstance(node)) {
    let root = normalizeNode(node.block)
    if (!root) return node
    realNode = root
  } else {
    realNode = node
  }

  if (!instance.dirs.has(realNode)) instance.dirs.set(realNode, [])
  const bindings = instance.dirs.get(realNode)!

  for (const directive of directives) {
    let [dir, source, arg, modifiers] = directive
    if (!dir) continue
    if (isFunction(dir)) {
      dir = {
        mounted: dir,
        updated: dir,
      } satisfies ObjectDirective
    }

    const binding: DirectiveBinding = {
      dir,
      instance,
      source,
      value: null, // set later
      oldValue: undefined,
      arg,
      modifiers,
    }
    bindings.push(binding)

    callDirectiveHook(realNode, binding, instance, 'created')

    // register source
    if (source) {
      if (dir.deep) {
        const deep = dir.deep === true ? undefined : dir.deep
        const baseSource = source
        source = () => traverse(baseSource(), deep)
      }
      renderEffect(source)
    }
  }

  return node

  function normalizeNode(block: Block | null) {
    if (!block) return

    if (isArray(block)) {
      __DEV__ &&
        warn(
          `Runtime directive used on component with non-element root node. ` +
            `The directives will not function as intended.`,
        )
      return
    }

    if (block instanceof Node || fragmentKey in block) {
      return block as Node
    } else {
      return normalizeNode(block.block)
    }
  }
}

export function invokeDirectiveHook(
  instance: ComponentInternalInstance | null,
  name: DirectiveHookName,
  nodes?: IterableIterator<Node>,
) {
  if (!instance) return
  nodes = nodes || instance.dirs.keys()
  for (const node of nodes) {
    const directives = instance.dirs.get(node) || []
    for (const binding of directives) {
      callDirectiveHook(node, binding, instance, name)
    }
  }
}

function callDirectiveHook(
  node: Node,
  binding: DirectiveBinding,
  instance: ComponentInternalInstance | null,
  name: DirectiveHookName,
) {
  if (name === 'beforeUpdate') binding.oldValue = binding.value
  const { dir } = binding
  const hook = dir[name]
  if (!hook) return

  const newValue = binding.source ? binding.source() : undefined
  binding.value = newValue
  // disable tracking inside all lifecycle hooks
  // since they can potentially be called inside effects.
  pauseTracking()
  callWithAsyncErrorHandling(hook, instance, VaporErrorCodes.DIRECTIVE_HOOK, [
    node,
    binding,
  ])
  resetTracking()
}
