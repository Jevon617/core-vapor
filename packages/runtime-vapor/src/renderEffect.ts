import {
  EffectFlags,
  ReactiveEffect,
  type SchedulerJob,
  SchedulerJobFlags,
} from '@vue/reactivity'
import { invokeArrayFns } from '@vue/shared'
import {
  type ComponentInternalInstance,
  getCurrentInstance,
  setCurrentInstance,
} from './component'
import { queueJob, queuePostRenderEffect } from './scheduler'
import { VaporErrorCodes, callWithAsyncErrorHandling } from './errorHandling'
import { invokeDirectiveHook } from './directives'

export function renderEffect(cb: () => void) {
  const instance = getCurrentInstance()

  let effect: ReactiveEffect

  const job: SchedulerJob = () => {
    if (!(effect.flags & EffectFlags.ACTIVE) || !effect.dirty) {
      return
    }

    if (instance?.isMounted && !instance.isUpdating) {
      instance.isUpdating = true

      const { bu, u, dirs } = instance
      // beforeUpdate hook
      if (bu) {
        invokeArrayFns(bu)
      }
      if (dirs) {
        invokeDirectiveHook(instance, 'beforeUpdate')
      }

      effect.run()

      queuePostRenderEffect(() => {
        instance.isUpdating = false
        if (dirs) {
          invokeDirectiveHook(instance, 'updated')
        }
        // updated hook
        if (u) {
          queuePostRenderEffect(u)
        }
      })
    } else {
      effect.run()
    }
  }

  effect = new ReactiveEffect(() => {
    const reset = instance && setCurrentInstance(instance)
    callWithAsyncErrorHandling(cb, instance, VaporErrorCodes.RENDER_FUNCTION)
    reset?.()
  })

  effect.scheduler = () => {
    if (instance) job.id = instance.uid
    queueJob(job)
  }

  effect.run()
}

export function firstEffect(
  instance: ComponentInternalInstance,
  fn: () => void,
) {
  const effect = new ReactiveEffect(fn)
  const job: SchedulerJob = () => effect.run()
  job.flags! |= SchedulerJobFlags.PRE
  job.id = instance.uid
  effect.scheduler = () => queueJob(job)
  effect.run()
}
