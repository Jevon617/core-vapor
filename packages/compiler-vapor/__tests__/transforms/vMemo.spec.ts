import { NodeTypes } from '@vue/compiler-dom'
import { IRNodeTypes } from '../../src'
import { getBaseTransformPreset } from '../../src/compile'
import { makeCompile } from './_utils'

const [nodeTransforms, directiveTransforms] = getBaseTransformPreset(true)
const compileWithMemo = makeCompile({
  nodeTransforms,
  directiveTransforms,
})

describe('compiler: v-memo', () => {
  test('on normal element', () => {
    const { ir, code, vaporHelpers } = compileWithMemo(
      `<div v-memo="[x]">
        {{ y }}
      </div>`,
    )
    expect(code).toMatchSnapshot()
    expect(vaporHelpers).contains('createMemo')
    expect(ir.block.effect).lengthOf(0)
    expect(ir.block.operation).toMatchObject([
      {
        type: IRNodeTypes.MEMO,
        id: 0,
        memo: {
          type: NodeTypes.SIMPLE_EXPRESSION,
          content: '[x]',
        },
        render: {
          type: IRNodeTypes.BLOCK,
          dynamic: {
            children: [{ template: 0 }],
          },
        },
        level: 0,
      },
    ])
  })

  test('nested v-memo on normal element', () => {
    const { ir, code, vaporHelpers } = compileWithMemo(
      `<div v-memo="[x]">
        {{ y }}
        <div v-memo="[z]">
          {{ k }}
        </div>
      </div>`,
    )
    expect(code).toMatchSnapshot()
    expect(vaporHelpers).contains('createMemo')
    expect(ir.block.effect).lengthOf(0)
    expect(ir.block.operation).toMatchObject([
      {
        type: IRNodeTypes.MEMO,
        id: 0,
        memo: {
          type: NodeTypes.SIMPLE_EXPRESSION,
          content: '[x]',
        },
        render: {
          type: IRNodeTypes.BLOCK,
          dynamic: {
            children: [{ template: 0 }],
          },
        },
        level: 0,
      },
    ])
    expect((ir.block.operation[0] as any).render.operation[1]).toMatchObject({
      type: IRNodeTypes.MEMO,
      id: 3,
      memo: {
        type: NodeTypes.SIMPLE_EXPRESSION,
        content: '[z]',
      },
      render: {
        type: IRNodeTypes.BLOCK,
        dynamic: {
          children: [{ template: 0 }],
        },
      },
      level: 1,
    })
  })

  test.todo('on component')
  test.todo('on slot outlet')
  test.todo('with v-if/else')
  test.todo('with v-for')
})
