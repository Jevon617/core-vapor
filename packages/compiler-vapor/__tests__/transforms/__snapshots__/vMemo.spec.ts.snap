// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`compiler: v-memo > basic 1`] = `
"import { setText as _setText, createMemo as _createMemo, template as _template } from 'vue/vapor';
const t0 = _template("<div></div>")

export function render(_ctx) {
  const n0 = _createMemo(() => ([_ctx.x]), (_cache) => {
    const n2 = t0()
    _setText(n2, _ctx.y)
    return n2
  }, 0)
  return n0
}"
`;

exports[`compiler: v-memo > nested v-memo on normal element 1`] = `
"import { createTextNode as _createTextNode, setText as _setText, createMemo as _createMemo, insert as _insert, template as _template } from 'vue/vapor';
const t0 = _template("<div></div>")

export function render(_ctx) {
  const n0 = _createMemo(() => ([_ctx.x]), (_cache) => {
    const n6 = t0()
    const n2 = _createTextNode([_ctx.y, " "])
    const n3 = _createMemo(() => ([_ctx.z]), (_cache) => {
      const n5 = t0()
      _setText(n5, _ctx.k)
      return n5
    }, 1, _cache)
    _insert([n2, n3], n6)
    return n6
  }, 0)
  return n0
}"
`;

exports[`compiler: v-memo > on normal element 1`] = `
"import { setText as _setText, createMemo as _createMemo, template as _template } from 'vue/vapor';
const t0 = _template("<div></div>")

export function render(_ctx) {
  const n0 = _createMemo(() => ([_ctx.x]), (_cache) => {
    const n2 = t0()
    _setText(n2, _ctx.y)
    return n2
  }, 0)
  return n0
}"
`;
